config.json RequestOptions mapping no longer working in Vertx 3.9.5 and Jackson 2.10.4 or 2.11.3

### Questions

Do not use this issue tracker to ask questions, instead use one of these [channels](https://vertx.io/community/). Questions will likely be closed without notice.

### Version

Vertx 3.9.5 and Jackson 2.10.4 or 2.11.3

### Context

I encountered an exception which looks suspicious while upgrading from Vertx 3.6.3 and Jackson 2.9.3

### Do you have a reproducer?

A reproducer is a simple project hosted on GitHub (or another forge supporting git clone operation) that has a build file that can be executed to reproduce the issue.

Reproducers are very helpful for contributors and will likely help them fixing your bug faster.

* Link to github project/gist

### Steps to reproduce

1. create unit tests as follows and run:

    public void checkRequestOptionsDeserialize() throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        TypeFactory typeFactory = mapper.getTypeFactory();
        JavaType javaType =  typeFactory.constructType(RequestOptions.class);
        RequestOptions requestOptions = new RequestOptions();
        String json = mapper.writeValueAsString(requestOptions);
        System.out.println(json);

        WebClientApiConfigurationTest deserializedObject = mapper.readValue(json, javaType);
    }

Error results:


Conflicting setter definitions for property "all": io.vertx.core.MultiMap#setAll(1 params) vs io.vertx.core.MultiMap#setAll(1 params)
 at [Source: (String)"{"host":"localhost","port":80,"ssl":null,"uri":"","headers":null}"; line: 1, column: 1]
com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "all": io.vertx.core.MultiMap#setAll(1 params) vs io.vertx.core.MultiMap#setAll(1 params)
 at [Source: (String)"{"host":"localhost","port":80,"ssl":null,"uri":"","headers":null}"; line: 1, column: 1]
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:309)


2) 2nd test

    @Test
    public void checkRequestOptionssnsCanDeserialize() throws JsonProcessingException {
        ObjectMapper mapper = new ObjectMapper();
        TypeFactory typeFactory = mapper.getTypeFactory();
        JavaType javaType =  typeFactory.constructType(RequestOptions.class);

        AtomicReference<Throwable> causeRef = new AtomicReference<>();
        if (! mapper.canDeserialize(javaType, causeRef)) {
            throw new RuntimeException(causeRef.get());
        }
    }

results:

com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "all": io.vertx.core.MultiMap#setAll(1 params) vs io.vertx.core.MultiMap#setAll(1 params)
java.lang.RuntimeException: com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "all": io.vertx.core.MultiMap#setAll(1 params) vs io.vertx.core.MultiMap#setAll(1 params)
	at com.nbcuni.oa.mam.workorder.jackson.JacksonOverloadedSettersTest.checkRequestOptionssnsCanDeserialize(JacksonOverloadedSettersTest.java:154)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.runTestClass(JUnitTestClassExecutor.java:110)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:58)
	at org.gradle.api.internal.tasks.testing.junit.JUnitTestClassExecutor.execute(JUnitTestClassExecutor.java:38)
	at org.gradle.api.internal.tasks.testing.junit.AbstractJUnitTestClassProcessor.processTestClass(AbstractJUnitTestClassProcessor.java:62)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.processTestClass(SuiteTestClassProcessor.java:51)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.processTestClass(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.processTestClass(TestWorker.java:119)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:182)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:164)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:414)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
	at java.lang.Thread.run(Thread.java:748)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Conflicting setter definitions for property "all": io.vertx.core.MultiMap#setAll(1 params) vs io.vertx.core.MultiMap#setAll(1 params)


### Extra

* This may be related to a similar issue reported here:

https://github.com/FasterXML/jackson-databind/issues/2757

--- a/src/main/asciidoc/dataobjects.adoc
+++ b/src/main/asciidoc/dataobjects.adoc
@@ -1971,9 +1971,6 @@ Set proxy username.
 [frame="topbot"]
 |===
 ^|Name | Type ^| Description
-|[[headers]]`@headers`|`String`|+++
-Add a request header.
-+++
 |[[host]]`@host`|`String`|+++
 Set the host name to be used by the client request.
 +++

--- a/src/main/generated/io/vertx/core/http/RequestOptionsConverter.java
+++ b/src/main/generated/io/vertx/core/http/RequestOptionsConverter.java
@@ -14,14 +14,6 @@ public class RequestOptionsConverter {
   public static void fromJson(Iterable<java.util.Map.Entry<String, Object>> json, RequestOptions obj) {
     for (java.util.Map.Entry<String, Object> member : json) {
       switch (member.getKey()) {
-        case "headers":
-          if (member.getValue() instanceof JsonObject) {
-            ((Iterable<java.util.Map.Entry<String, Object>>)member.getValue()).forEach(entry -> {
-              if (entry.getValue() instanceof String)
-                obj.addHeader(entry.getKey(), (String)entry.getValue());
-            });
-          }
-          break;
         case "host":
           if (member.getValue() instanceof String) {
             obj.setHost((String)member.getValue());

--- a/src/main/java/io/vertx/core/http/RequestOptions.java
+++ b/src/main/java/io/vertx/core/http/RequestOptions.java
@@ -16,6 +16,8 @@ import io.vertx.codegen.annotations.GenIgnore;
 import io.vertx.core.MultiMap;
 import io.vertx.core.json.JsonObject;
 
+import java.util.List;
+import java.util.Map;
 import java.util.Objects;
 
 /**
@@ -81,7 +83,23 @@ public class RequestOptions {
    * @param json the JSON
    */
   public RequestOptions(JsonObject json) {
+    this();
     RequestOptionsConverter.fromJson(json, this);
+    JsonObject headers = json.getJsonObject("headers");
+    if (headers != null) {
+      for (Map.Entry<String, Object> entry : headers) {
+        Object value = entry.getValue();
+        if (value instanceof String) {
+          this.addHeader(entry.getKey(), (String) value);
+        } else if (value instanceof Iterable) {
+          for (Object subValue : ((Iterable<?>) value)) {
+            if (subValue instanceof String) {
+              this.addHeader(entry.getKey(), (String) subValue);
+            }
+          }
+        }
+      }
+    }
   }
 
   /**
@@ -165,6 +183,7 @@ public class RequestOptions {
    * @param value  the header value
    * @return a reference to this, so the API can be used fluently
    */
+  @GenIgnore
   public RequestOptions addHeader(String key, String value) {
     checkHeaders();
     Objects.requireNonNull(key, "no null key accepted");
@@ -204,6 +223,18 @@ public class RequestOptions {
   public JsonObject toJson() {
     JsonObject json = new JsonObject();
     RequestOptionsConverter.toJson(this, json);
+    if (this.headers != null) {
+      JsonObject headers = new JsonObject();
+      for (String name : this.headers.names()) {
+        List<String> values = this.headers.getAll(name);
+        if (values.size() == 1) {
+          headers.put(name, values.iterator().next());
+        } else {
+          headers.put(name, values);
+        }
+      }
+      json.put("headers", headers);
+    }
     return json;
   }
 }

--- a/src/test/java/io/vertx/core/http/RequestOptionsTest.java
+++ b/src/test/java/io/vertx/core/http/RequestOptionsTest.java
@@ -1,0 +1,84 @@
+/*
+ * Copyright (c) 2011-2021 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.http;
+
+import io.vertx.core.MultiMap;
+import io.vertx.core.json.JsonArray;
+import io.vertx.core.json.JsonObject;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class RequestOptionsTest {
+
+  @Test
+  public void testDefaults() {
+    RequestOptions options = new RequestOptions();
+    assertEquals(RequestOptions.DEFAULT_HOST, options.getHost());
+    assertEquals(RequestOptions.DEFAULT_PORT, options.getPort());
+    assertEquals(RequestOptions.DEFAULT_SSL, options.isSsl());
+    assertEquals(RequestOptions.DEFAULT_URI, options.getURI());
+  }
+
+  @Test
+  public void testCopy() {
+    RequestOptions options = new RequestOptions()
+      .setPort(8443)
+      .setSsl(true);
+    RequestOptions copy = new RequestOptions(options);
+    assertEquals(options.getPort(), copy.getPort());
+    assertEquals(options.isSsl(), copy.isSsl());
+  }
+
+  @Test
+  public void testToJson() {
+    RequestOptions options = new RequestOptions()
+      .setPort(8443)
+      .setSsl(true)
+      .addHeader("key", "value")
+      .addHeader("foo", "bar")
+      .addHeader("foo", "baz");
+    JsonObject expected = new JsonObject()
+      .put("host", RequestOptions.DEFAULT_HOST)
+      .put("uri", RequestOptions.DEFAULT_URI)
+      .put("port", 8443)
+      .put("ssl", true)
+      .put("headers", new JsonObject()
+        .put("key", "value")
+        .put("foo", new JsonArray().add("bar").add("baz"))
+      );
+    assertEquals(expected, options.toJson());
+  }
+
+  @Test
+  public void testFromJson() {
+    JsonObject json = new JsonObject()
+      .put("method", "PUT")
+      .put("port", 8443)
+      .put("ssl", true)
+      .put("headers", new JsonObject()
+        .put("key", "value")
+        .put("foo", new JsonArray().add("bar").add("baz"))
+      );
+    RequestOptions options = new RequestOptions(json);
+    assertEquals(8443, options.getPort());
+    assertTrue(options.isSsl());
+    MultiMap headers = options.getHeaders();
+    assertEquals(headers.toString(), 2, headers.size());
+    assertEquals(Collections.singletonList("value"), headers.getAll("key"));
+    assertEquals(Arrays.asList("bar", "baz"), headers.getAll("foo"));
+  }
+}

